k8s/anthos/machines.go

```diff
@@ -0,0 +1,76 @@
+package anthos
+
+import (
+	"context"
+	"encoding/json"
+
+	v1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/deprecated/v1alpha1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	machineResource = schema.GroupVersionResource{Group: "cluster.k8s.io", Version: "v1alpha1", Resource: "machines"}
+)
+
+// MachinesOps is an interface to perform k8s machines operations
+type MachineOps interface {
+	// ListMachines lists all machines in kubernetes cluster
+	ListMachines(ctx context.Context) (*v1alpha1.MachineList, error)
+	// GetMachine returns a machine for the given name
+	GetMachine(ctx context.Context, name string) (*v1alpha1.Machine, error)
+	// DeleteMachine delete machine for given name
+	DeleteMachine(ctx context.Context, name string) error
+}
+
+// ListMachines lists all machines in kubernetes cluster
+func (c *Client) ListMachines(ctx context.Context) (*v1alpha1.MachineList, error) {
+	if err := c.initClient(); err != nil {
+		return nil, err
+	}
+
+	result, err := c.dynamicClient.Resource(machineResource).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	jsonData, err := result.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	machineList := &v1alpha1.MachineList{}
+	if err := json.Unmarshal(jsonData, machineList); err != nil {
```

Comment:
I think we can directly Unmarshal the result into `machineList` without converting it into JSON.

```go
result, err := c.dynamicClient.Resource(machineResource).List(ctx, metav1.ListOptions{})
if err != nil {
	return nil, err
}
machineList := &v1alpha1.MachineList{}
if err := json.Unmarshal(result, machineList); err != nil {
	return nil, err
}
```

Reply:
@kphalgun-px  here result will be  [`*unstructured.UnstructuredList` ](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured)  so MarshalJSON() has to be used. How will you parse Unstructured data to json ?

Reply:
My bad. Thanks I thought results will be in byte slice format.